### PR TITLE
podgc metrics should count all pod deletion behaviors

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Literally, `force_delete_pods_total ` and `force_delete_pod_errors_total ` metrics  should count all the behaviors of the gc controller that forcefully delete pods.

The `reason`  label indicates the reason why the pod was forcibly deleted by GC

#### Does this PR introduce a user-facing change?

```release-note
`force_delete_pods_total ` and `force_delete_pod_errors_total ` metrics count all pod deletion behaviors.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2268
```

### CODE REVIEW

File: pkg/controller/podgc/gc_controller_test.go [hunk 1]
```diff
@@ -672,15 +672,15 @@ func verifyDeletedAndPatchedPods(t *testing.T, client *fake.Clientset, wantDelet
...
 	if err != nil {
 		t.Errorf("Error getting actualDeletingPodsTotal")
 	}
 	if actualDeletingPodsTotal != float64(inputDeletingPodsTotal) {
 		t.Errorf("Expected desiredDeletingPodsTotal to be %d, got %v", inputDeletingPodsTotal, actualDeletingPodsTotal)
 	}
 
-	actualDeletingPodsErrorTotal, err := metricstestutil.GetCounterMetricValue(deletingPodsErrorTotal.WithLabelValues())
+	actualDeletingPodsErrorTotal, err := metricstestutil.GetCounterMetricValue(deletingPodsErrorTotal.WithLabelValues("out-of-service"))
 	if err != nil {
```
Comment (xing-yang) [inline pkg/controller/podgc/gc_controller_test.go:684]: Can you add unit tests for all the other labels?

Comment (carlory) [reply pkg/controller/podgc/gc_controller_test.go:693]: added.

File: pkg/controller/podgc/gc_controller.go [hunk 2]
```diff
@@ -179,11 +179,11 @@ func (gcc *PodGCController) gcTerminating(ctx context.Context, pods []*v1.Pod) {
 		wait.Add(1)
 		go func(pod *v1.Pod) {
 			defer wait.Done()
-			deletingPodsTotal.WithLabelValues().Inc()
+			deletingPodsTotal.WithLabelValues("out-of-service").Inc()
```
Comment (gnufied) [inline pkg/controller/podgc/gc_controller.go:182]: Can we move these dimensions such as `out-of-service`, `terminated` etc in well defined constants, so as it gets easier to reuse them without potentially having typos in the code?

File: pkg/controller/podgc/gc_controller.go [hunk 3]
```diff
@@ -213,9 +213,11 @@ func (gcc *PodGCController) gcTerminated(ctx context.Context, pods []*v1.Pod) {
 		wait.Add(1)
 		go func(pod *v1.Pod) {
 			defer wait.Done()
+			deletingPodsTotal.WithLabelValues("terminated").Inc()
```
Comment (gnufied) [inline pkg/controller/podgc/gc_controller.go:216]: I am thinking, if we should emit this metric before or after the completion of delete operation. As it stands, we emit when we start deleting.

Also, I think we should add namespace to these metrics, so as if we force deleted pods from a particular namespace, a k8s admin can investigate why that happened. @msau42 what do you think?

File: pkg/controller/podgc/gc_controller.go (hunk 2 above)
Comment (carlory) [reply pkg/controller/podgc/gc_controller.go]: @gnufied updated.

File: pkg/controller/podgc/gc_controller.go (hunk 3 above)
Comment (carlory) [reply pkg/controller/podgc/gc_controller.go]: Waiting for a feedback from @msau42

Comment (msau42) [reply pkg/controller/podgc/gc_controller.go]: namespace would be good to add. I have a slight preference to add the metric after completion. Do we care about deletion errors?

Comment (carlory) [reply pkg/controller/podgc/gc_controller.go]: @xing-yang what do you think？

**Comment by carlory** [comment]
@xing-yang @msau42 @gnufied the namespace label is added.

File: pkg/controller/podgc/gc_controller_test.go [hunk 4]
```diff
@@ -669,18 +679,18 @@ func verifyDeletedAndPatchedPods(t *testing.T, client *fake.Clientset, wantDelet
 	}
 }
 
-func testDeletingPodsMetrics(t *testing.T, inputDeletingPodsTotal int) {
+func testDeletingPodsMetrics(t *testing.T, total int, reason string) {
 	t.Helper()
 
-	actualDeletingPodsTotal, err := metricstestutil.GetCounterMetricValue(deletingPodsTotal.WithLabelValues())
+	actualDeletingPodsTotal, err := metricstestutil.GetCounterMetricValue(metrics.DeletingPodsTotal.WithLabelValues("", reason))
```
Comment (xing-yang) [inline pkg/controller/podgc/gc_controller_test.go:685]: What is the namespace for these pods?

Comment (carlory) [reply pkg/controller/podgc/gc_controller_test.go]: @xing-yang It's fixed in https://github.com/kubernetes/kubernetes/compare/04ac55e68117dff43adaaf3be47e39578e44dadc..322da7c1aae34496fda5ce5dd96966706d6513b4

**Review by soltysh — APPROVED** [review]
/approve

**Pull request merged**